homeassistant/components/logbook/queries.py

```diff
@@ -292,27 +296,67 @@ def _select_events_without_states(
     )
 
 
-def _select_states(start_day: dt, end_day: dt) -> Select:
+def _states_query_for_context_id(start_day: dt, end_day: dt, context_id: str) -> Query:
+    return _apply_states_filters(_select_states(), start_day, end_day).where(
+        States.context_id == context_id
+    )
+
+
+def _states_query_for_entitiy_id(start_day: dt, end_day: dt, entity_id: str) -> Query:
+    return _apply_states_filters(
+        _apply_entities_hints(_select_states()), start_day, end_day
+    ).where(States.entity_id == entity_id)
+
+
+def _states_query_for_entitiy_ids(
+    start_day: dt, end_day: dt, entity_ids: list[str]
+) -> Query:
+    return _apply_states_filters(
+        _apply_entities_hints(_select_states()), start_day, end_day
+    ).where(States.entity_id.in_(entity_ids))
+
+
+def _states_query_for_all(start_day: dt, end_day: dt) -> Query:
+    return _apply_states_filters(_apply_all_hints(_select_states()), start_day, end_day)
+
+
+def _select_states() -> Select:
     """Generate a states select that formats the states table as event rows."""
-    old_state = aliased(States, name="old_state")
+    return select(
+        literal(value=None, type_=sqlalchemy.Text).label("event_id"),
+        literal(value=EVENT_STATE_CHANGED, type_=sqlalchemy.String).label("event_type"),
+        literal(value=None, type_=sqlalchemy.Text).label("event_data"),
+        States.last_updated.label("time_fired"),
+        States.context_id.label("context_id"),
+        States.context_user_id.label("context_user_id"),
+        States.context_parent_id.label("context_parent_id"),
+        literal(value=None, type_=sqlalchemy.Text).label("shared_data"),
+        *STATE_COLUMNS,
+        NOT_CONTEXT_ONLY,
+    )
+
+
+def _apply_all_hints(query: Query) -> Query:
```

Comment:
I had to refactor the queries a bit to be able to insert the two hints here.

There should be no other functional change here than the two `.with_hint` lines